File: POCs/02_idea_progression_react/src/session/shared/session-service.ts
import { DateTime } from "luxon";
import { Exercise, SessionVM, SetTimeMetrics, SetType, SetWeightMetrics, WorkoutPlan } from "./session.model";

export function setupFakeExercises(){
    return [
        { 
            metricType: 'weight',
            name: "Barbell Bench Press", 
            targetSets: [
                {setType: "Warm-up", target: {weight:20, reps: 10} },
                {setType: "Warm-up", target: {weight: 20, reps: 10}},
                {setType: "Set", target: {weight: 55, reps: 8} },
                {setType: "Set", target: {weight: 55, reps: 8}},
                {setType: "Set", target: {weight: 55, reps: 8}}
            ]
        } as Exercise<SetWeightMetrics>,
        { 
            metricType: 'weight',
            name: "Dumbbell Bench Press", 
            targetSets: [
                {setType: "Set", target: {weight: 55, reps: 8}},
                {setType: "Set", target: {weight: 55, reps: 8}},
                {setType: "Set", target: {weight: 55, reps: 8}}
            ]
        } as Exercise<SetWeightMetrics>,
        { 
            metricType: 'weight',
            name: "Triceps Pushdown", 
            targetSets: [
                {setType: "Set", target: {weight: 55, reps: 8}},
                {setType: "Set", target: {weight: 55, reps: 8}},
                {setType: "Set", target: {weight: 55, reps: 8}},
            ]
        } as Exercise<SetWeightMetrics>,
        { 
            metricType: 'time',
            name: "Plank", 
            targetSets: [
                {setType: "Set", target: {timeMilliseconds: 60000}},
                {setType: "Set", target: {timeMilliseconds: 60000}}
            ]
        } as Exercise<SetTimeMetrics>,
    ];
}


export function setupFakeSession(){
    let workoutPlan: WorkoutPlan = {
        id: "1",
        name: "Push Day",
        exercises: setupFakeExercises()
    }

    let session: SessionVM = {
        workoutPlan: workoutPlan,
        workoutStart: DateTime.now(),
        exercises: workoutPlan.exercises.map(workoutExercise => ({
            exercise: workoutExercise,
            sets: workoutExercise.targetSets.map(targetSet => ({
                setType: targetSet.setType,
                targetMetrics: targetSet.target,
                actualMetrics: null,
                completed: false,
                isEditing: false
            })),
            completed: false
        }))
    };

    return session;
}

File: POCs/04_Blazor_Fluxor/GymTrackerBlazorFluxorPOC/Session/SessionService.cs
namespace GymTrackerBlazorFluxorPOC.Session;

public class SessionService
{
    //public SessionVM FetchSession()


    public List<Exercise<SetMetrics>> BuildExercises()
    {
        return new List<Exercise<SetMetrics>>
        {
            new Exercise<SetMetrics>{Id = 1, Name = "Incline Bench Press", TargetSets=null},
            new Exercise<SetMetrics>{Id = 2, Name = "Incline Dumbell Bench Press", TargetSets=null},
        };
    }

    public List<IExercise<SetMetrics>> SetupFakeExercises()
    {
        return new List<IExercise<SetMetrics>>
        {
            new Exercise<SetMetrics>{
                Name = "Barbell Bench Press",
                TargetSets = new List<ExerciseTargetSet<SetMetrics>>
                {
                    new ExerciseTargetSet<SetMetrics>
                    {
                        SetType = "Warm Up",
                        Target = new SetWeightMetrics{Weight = 20, Reps = 10}
                    }
                }
            },
            new Exercise<SetMetrics>{
                Name = "Dumbell Bench Press",
                TargetSets = new List<ExerciseTargetSet<SetMetrics>>
                {
                    new ExerciseTargetSet<SetMetrics>
                    {
                        SetType = "Warm Up",
                        Target = new SetWeightMetrics{Weight = 20, Reps = 10}
                    }
                }
            },
            new Exercise<SetMetrics>{
                Name = "Triceps Pushdown",
                TargetSets = new List<ExerciseTargetSet<SetMetrics>>
                {
                    new ExerciseTargetSet<SetMetrics>
                    {
                        SetType = "Warm Up",
                        Target = new SetWeightMetrics{Weight = 20, Reps = 10}
                    }
                }
            }
        };
    }

    public SessionVM? CurrentSession { get; private set; }

    public SessionVM SetupFakeSession()
    {
        var workoutPlan = new WorkoutPlan
        {
            Id = 1,
            Name = "Push Day",
            Exercises = SetupFakeExercises()
        };

        var session = new SessionVM
        {
            WorkoutPlan = workoutPlan,
            WorkoutStart = DateTime.Now,
            Exercises = workoutPlan.Exercises.Select(workoutExercise => new ExerciseVM<SetMetrics>
            {
                Id = Guid.NewGuid(),
                Exercise = workoutExercise,
                Sets = workoutExercise.TargetSets.Select(targetSet => new ExerciseSetVM<SetMetrics> { SetType = targetSet.SetType, TargetMetrics = targetSet.Target, ActualMetrics = null })
            }).ToList()
        };

        CurrentSession = session;

        return session;
    }

}
